In [1]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)

1. Найти с помощью NumPy SVD для матрицы

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$


In [10]:
A = np.array([[1, 2, 3],
              [0, 0, 5],
              [3,-4, 2],
              [1, 6, 5],
              [0, 1, 0]])
print(f'Матрица A:\n{A}')

Матрица A:
[[ 1  2  3]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]


In [36]:
U, s, W = np.linalg.svd(A)

In [37]:
D[np.diag_indices((min(A.shape)))] = s
print(f'Матрица D:\n{D}')

Матрица D:
[[9.43 0.   0.   0.   0.  ]
 [0.   6.11 0.   0.   0.  ]
 [0.   0.   2.16 0.   0.  ]
 [0.   0.   0.   0.   0.  ]
 [0.   0.   0.   0.   0.  ]]


In [38]:
print(f'Матрица U:\n{U}')

Матрица U:
[[ 0.39 -0.11 -0.18 -0.88 -0.15]
 [ 0.4  -0.46  0.78  0.05  0.13]
 [-0.08 -0.85 -0.48  0.16  0.07]
 [ 0.82  0.18 -0.33  0.42 -0.07]
 [ 0.07  0.12 -0.12 -0.12  0.98]]


In [44]:
print(f'Матрица V:\n{W.T}')

Матрица V:
[[ 0.1  -0.41 -0.91]
 [ 0.65  0.72 -0.25]
 [ 0.75 -0.56  0.34]]


In [56]:
# Проведем проверку
np.dot(np.dot(U, D)[:, 0:3], W)

array([[ 1.,  2.,  3.],
       [ 0.,  0.,  5.],
       [ 3., -4.,  2.],
       [ 1.,  6.,  5.],
       [ 0.,  1., -0.]])

---

2. Для матрицы из предыдущего задания найти:

    а) евклидову норму;
    
    б) норму Фробениуса.

$\left \| A \right \|_{E}=\mu_{1}$

$\left \| A \right \|_{E}=9.43$

$\left \| A \right \|_{F}=\sqrt{\sum_{k=1}^{r}\mu_{k}^{2}}$

In [58]:
s.sum()

17.707891073002948

$\left \| A \right \|_{F}=17.71$